### ISSUE TITLE
Fault tolerance with k8s

### ISSUE THREAD

**Issue by jianzi123** [pytorch/pytorch#136312]
### 🚀 The feature, motivation and pitch

I am working on training fault tolerance. 
We want to restart only one training node when there is a hardware failure, but the existing design does not allow the agent to restart the training process. I hope that it can be realized that the agent can exit when an error occurs, and other non-faulty training node agents can survive

### Alternatives

1. Write the number of restarts to etcd/tcp_store
2. Add configuration parameters. If a training process exits with error, the agent exits

**Comment by kwen2501** [comment]
> the existing design does not allow the agent exit and will restart the training process

Yeah, the current implementation is a bit harsh, in that the watchdog will kill the process when detecting failure or timeout. 

As a first step, we can relax this and leave the process alive, which would allow the user to take control. 

Second step, we imagine that the user might want to kill the dead communicator -- a NCCL kernel may be hanging, etc. For this, we are experimenting with a new API, see https://github.com/pytorch/pytorch/pull/132291

**Comment by jianzi123** [comment]
@kwen2501 The second step seems to be a final solution with some architectural changes. We are currently trying to make some modifications to complete the first step, which allows the agent to exit and let the external system decide whether to continue training or not, so there is some flexibility to do things externally. Discuss it if necessary
